### PR TITLE
Add missing shortTimeLabel for text marks, support formatting for tex…

### PR DESCRIPTION
**Pull request by domoritz**
…t marks, simplify logic.

Follow up for #937

### CODE REVIEW

File: src/compile/Model.ts [hunk 1]
```diff
@@ -233,48 +233,88 @@ export class Model {
     return 30;
   }
 
-  /** @return an object with format and formatType properties. */
+  /**
+   * Returns an object with format and formatType properties.
```
Comment (kanitw) [inline src/compile/Model.ts]: @param format explicitly specified format?

File: src/compile/Model.ts [hunk 2]
```diff
@@ -233,48 +233,88 @@ export class Model {
...
+      const filter = def.formatType + (def.format ? ':\'' + def.format + '\'' : '');
+      return {
+        text: {
+          template: '{{' + this.field(channel, {datum: true}) + ' | ' + filter + '}}'
         }
-        break;
+      };
+    }
+
+    if (def.formatType === 'number') {
```
Comment (kanitw) [inline src/compile/Model.ts]: Why setting it first and remove later?  Better not set it at all from the beginning.

File: src/compile/Model.ts [hunk 3]
```diff
@@ -233,48 +233,88 @@ export class Model {
...
+      case COLOR:
+      case SHAPE:
+      case SIZE:
+        const legend = fieldDef.legend;
+        return (typeof legend !== 'boolean' ? legend.shortTimeLabels : false);
+      case TEXT:
+        return this.config().mark.shortTimeLabels;
+      case LABEL:
+        // TODO(#897): implement when we have label
+      case DETAIL:
```
Comment (kanitw) [inline src/compile/Model.ts]: Detail doesn't have scale and thus shouldn't be called at all.  No need to add this condition.

File: src/compile/Model.ts (hunk 2 above)
Comment (domoritz) [inline src/compile/Model.ts]: See the check in :273

File: src/compile/Model.ts (hunk 3 above)
Comment (domoritz) [inline src/compile/Model.ts]: I like to have exhaustive switch statements.

File: src/compile/Model.ts [hunk 4]
```diff
@@ -233,48 +233,88 @@ export class Model {
     return 30;
   }
 
-  /** @return an object with format and formatType properties. */
+  /**
+   * Returns an object with format and formatType properties.
+   */
   public formatMixins(channel: Channel, format: string) {
```
Comment (kanitw) [inline src/compile/Model.ts]: Now that I add util.ts -- should this be moved to util.ts?   

One possible way to make distinction is things `util.ts` might either contains mixins or slightly complicated logic, while methods in model could be something more like helper?

File: src/compile/Model.ts [hunk 5]
```diff
@@ -233,48 +233,88 @@ export class Model {
...
+    if (def.formatType === 'number') {
+      // no need to set vega default
+      delete def.formatType;
     }
 
     return def;
   }
 
-  /** returns the time format used for axis labels for a time unit */
+  private isAbbreviated(channel: Channel, fieldDef: FieldDef) {
```
Comment (kanitw) [inline src/compile/Model.ts]: util.ts?

File: src/compile/legend.ts [hunk 6]
```diff
@@ -81,7 +81,7 @@ export function formatMixins(model: Model, channel: Channel) {
   }
 
   const legend = fieldDef.legend;
-  return model.formatMixins(channel, (typeof legend !== 'boolean' && legend.format) || undefined);
+  return model.formatMixins(channel, (typeof legend !== 'boolean' && legend.format));
```
Comment (kanitw) [inline src/compile/legend.ts]: formatMixins accept string as second parameter  
 should it be 
`typeof legend !== 'boolean' ? legend.format : undefined`  

although it pretty much gives same result.

File: test/compile/mark-text.test.ts [hunk 7]
```diff
@@ -23,7 +23,43 @@ describe('Mark: Text', function() {
...
+    let spec = {
+      "mark": "text",
+      "encoding": {
+        "text": {"field": "foo", "type": "temporal"}
+      }
+    };
+    const model = parseModel(spec);
+    const props = text.properties(model);
+
+    it('should use dtae template', function() {
```
Comment (kanitw) [inline test/compile/mark-text.test.ts]: date

File: test/compile/mark-text.test.ts [hunk 8]
```diff
@@ -23,7 +23,43 @@ describe('Mark: Text', function() {
     it('should have placeholder text', function() {
       assert.deepEqual(props.text, {value: "Abc"});
     });
-  })
+  });
+
+  describe('with quantitative and format', function() {
+    let spec = {
```
Comment (kanitw) [inline test/compile/mark-text.test.ts]: why `let` if you didn't modify it?

File: test/compile/mark-text.test.ts [hunk 9]
```diff
@@ -23,7 +23,43 @@ describe('Mark: Text', function() {
...
+    const model = parseModel(spec);
+    const props = text.properties(model);
+
+    it('should use number template', function() {
+      assert.deepEqual(props.text, {template: '{{datum.foo | number:\'d\'}}'});
+    });
+  });
+
+  describe('with temporal', function() {
+    let spec = {
```
Comment (kanitw) [inline test/compile/mark-text.test.ts]: why `let`?

File: src/compile/Model.ts [hunk 10]
```diff
@@ -233,48 +233,88 @@ export class Model {
...
     if (format !== undefined) {
       def.format = format;
-      return def;
+    } else {
+      switch (fieldDef.type) {
+        case QUANTITATIVE:
+          def.format = this.numberFormat(channel);
+          break;
+        case TEMPORAL:
+          const f = this.timeFormat(channel);
```
Comment (kanitw) [inline src/compile/Model.ts]: `def.format = this.timeFormat(channel) || this.config().timeFormat`
is more concise?

File: src/compile/mark-text.ts [hunk 11]
```diff
@@ -72,18 +73,9 @@ export namespace text {
...
-        p.text = {
-          template: '{{' + model.field(TEXT, { datum: true }) +
-          ' | number:\'' + numberFormat + '\'}}'
-        };
-      } else {
-        p.text = { field: model.field(TEXT) };
-      }
+      p.text = { field: model.field(TEXT) };
+      const format = model.config().mark.format;
+      extend(p, model.formatMixins(TEXT, format));
```
Comment (kanitw) [inline src/compile/mark-text.ts]: We generally avoid overriding and `p.text` is overridden for `Q` and `T`.

Better split this by `fieldDef.type`.   (And just call `formatMixins` for `Q` and `T`)

File: src/compile/Model.ts (hunk 2 above)
Comment (kanitw) [inline src/compile/Model.ts]: Better have 

`if ( contains([QUANTITATIVE, TEMPORAL], fieldDef.type) {`  

surrounds all logic in this function.  (`format` should be dropped for nominal and ordinal anyway!)

and do
```
if (fieldDef.type === TEMPORAL) {
   def.formatType = 'time'; 
}
```

File: src/compile/Model.ts (hunk 5 above)
Comment (domoritz) [inline src/compile/Model.ts]: I moved everything

File: src/compile/legend.ts (hunk 6 above)
Comment (domoritz) [inline src/compile/legend.ts]: done

File: test/compile/mark-text.test.ts (hunk 7 above)
Comment (domoritz) [inline test/compile/mark-text.test.ts]: my atom has become extremely sluggish recently

File: test/compile/mark-text.test.ts (hunk 8 above)
Comment (domoritz) [inline test/compile/mark-text.test.ts]: just copied

File: src/compile/Model.ts (hunk 10 above)
Comment (domoritz) [inline src/compile/Model.ts]: Done

File: src/compile/mark-text.ts (hunk 11 above)
Comment (domoritz) [inline src/compile/mark-text.ts]: What if the format is specified for ordinal?

File: src/compile/util.ts [hunk 12]
```diff
@@ -52,3 +55,84 @@ export function applyMarkConfig(marksProperties, model: Model, propsList: string
...
+
+    const filter = def.formatType + (def.format ? ':\'' + def.format + '\'' : '');
+    return {
+      text: {
+        template: '{{' + model.field(channel, { datum: true }) + ' | ' + filter + '}}'
+      }
+    };
+  }
+
+  if (def.formatType === 'number') {
```
Comment (kanitw) [inline src/compile/util.ts]: Still have this add and delete...

Comment (domoritz) [inline src/compile/util.ts]: Still because of line 92. I can drop this but it makes the logic hard to change.

File: src/compile/util.ts [hunk 13]
```diff
@@ -52,3 +55,84 @@ export function applyMarkConfig(marksProperties, model: Model, propsList: string
...
+
+  if (format !== undefined) {
+    def.format = format;
+  } else {
+    switch (fieldDef.type) {
+      case QUANTITATIVE:
+        def.format = model.numberFormat(channel);
+        break;
+      case TEMPORAL:
+        def.format = timeFormat(model, channel) || model.config().timeFormat;
```
Comment (kanitw) [inline src/compile/util.ts]: I guess `timeFormat` is only used here -- remove `timeFormat()` and move the logic here?

File: src/compile/util.ts [hunk 14]
```diff
@@ -52,3 +55,84 @@ export function applyMarkConfig(marksProperties, model: Model, propsList: string
...
+      def.formatType = 'time';
+      break;
+  }
+
+  if (format !== undefined) {
+    def.format = format;
+  } else {
+    switch (fieldDef.type) {
+      case QUANTITATIVE:
+        def.format = model.numberFormat(channel);
```
Comment (kanitw) [inline src/compile/util.ts]: I guess `numberFormat` is only used here -- remove `model.numberFormat()` and move the logic here?

Comment (domoritz) [inline src/compile/util.ts]: done

File: src/compile/util.ts (hunk 13 above)
Comment (domoritz) [inline src/compile/util.ts]: it's being tested as a whole

**Comment by domoritz** [comment]
Done, now format only works for Q and T.

**Comment by kanitw** [comment]
:tada:

File: src/schema/config.marks.schema.ts [hunk 15]
```diff
@@ -204,9 +205,13 @@ export const markConfig = {
     // text-only & VL only
     format: {
       type: 'string',
-      default: '',  // auto
-      description: 'The formatting pattern for text value.'+
-                   'If not defined, this will be determined automatically. '
+      default: undefined,  // auto
+      description: 'The formatting pattern for text value. If not defined, this will be determined automatically. '
+    },
+    shortTimeLabels: {
```
Comment (kanitw) [inline src/schema/config.marks.schema.ts]: ah, should also add this to `config.md`!

**Pull request merged**